src/nunit.analyzers.tests/TestCaseSourceUsage/TestCaseSourceUsesStringAnalyzerTests.cs

```diff
@@ -547,5 +546,158 @@ public class InnerClass
+            yield return 2;
+            yield return 3;
+        }
+    }", additionalUsings: "using System.Collections.Generic;");
+
+            var expectedDiagnostic = ExpectedDiagnostic
+                .Create(AnalyzerIdentifiers.TestCaseSourceMismatchInNumberOfTestMethodParameters)
+                .WithMessage("The TestCaseSource provides '1' parameter(s), but the Test method expects '2' parameter(s)");
+            RoslynAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenParameterTypeOfTestDiffersFromTestCaseSource()
+        {
+            var testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@"
+    [TestFixture]
+    public class AnalyzeWhenParameterTypeOfTestDiffersFromTestCaseSource
+    {
+        [TestCaseSource(↓nameof(TestData))]
+        public void ShortName(string message)
+        {
+            Assert.That(message.Length, Is.GreaterThanOrEqualTo(0));
+        }
+
+        static IEnumerable<int> TestData()
+        {
+            yield return 1;
+            yield return 2;
+            yield return 3;
+        }
+    }", additionalUsings: "using System.Collections.Generic;");
+
+            var expectedDiagnostic = ExpectedDiagnostic
+                .Create(AnalyzerIdentifiers.TestCaseSourceMismatchWithTestMethodParameterType)
+                .WithMessage("The TestCaseSource provides type 'int', but the Test method expects type 'string' for parameter 'message'");
+            RoslynAssert.Diagnostics(analyzer, expectedDiagnostic, testCode);
+        }
+
+        [Test]
+        public void AnalyzeWhenNumberOfParametersOfTestIsNotEvidentFromTestSource()
+        {
+            var testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@"
+    [TestFixture]
+    public class AnalyzeWhenNumberOfParametersOfTestIsNotEvidentFromTestSource
+    {
+        [Explicit(""The code is wrong, but it is too complext for the analyzer to detect this."")]
+        [TestCaseSource(nameof(TestData))]
+        public void ShortName(int n)
+        {
+            Assert.That(n, Is.GreaterThanOrEqualTo(0));
+        }
+
+        static IEnumerable<object> TestData()
+        {
+            yield return new object[] { 1, 2 };
+            yield return new object[] { 2, 3 };
+        }
+    }", additionalUsings: "using System.Collections.Generic;");
+
+            // The TestCaseSource provides '>0' parameter(s), but the Test method expects '1' parameter(s)
+            // Analyzing the actual code inside the TestCaseSource method is beyond the scope of the analyzer.
+            RoslynAssert.Valid(analyzer, testCode);
+        }
+
+        [TestCase("IEnumerable", "object", "System.Collections")]
+        [TestCase("IEnumerable<object>", "object", "System.Collections.Generic")]
+        [TestCase("IEnumerable<TestCaseData>", "TestCaseData", "System.Collections.Generic")]
+        [TestCase("IEnumerable<int>", "int", "System.Collections.Generic")]
```

Comment:
I guess the type can be determined here, but I think it is fine to keep the case

Reply:
Correct.